data/external/svo_ranking.sql

```diff
@@ -0,0 +1,99 @@
+--  - We use min-max normalisation for the intuitiveness of page hits with values closer to
+--      0 meaning they are less popular compared to pages with values closer to 1.
+--  To define our thinking/logic for the SQL, we define the following:
+--      x:      weekly page hits for each page
+--      group:  all pages per week
+
+--  ------------------
+--  Structure of query
+--  ------------------
+--  1.  Get relevant columns, `cte_all`.
+--          Includes transforming integer column to timestamp
+--  2.  Create id column, `cte_id`.
+--          This involves identifying normal pages (as defined by pagePath) from 'smart_answers' and
+--          'search' (as defined by other criteria)
+--  3.  Compute counts of page hits by week, `cte_week_counts`.
+--          We do it by week so we can account for the differences in weekday and weekend traffic together.
+--  4.  Compute the max and min page hits for all pages for each week, `cte_normalise`
+--          This is so we can normalise each page's page hits for each week
+--  5.  Compute the min-max normalised page hits for each page per week.
+
+--  ----------
+--  References
+--  ----------
+--  Use `SAFE_DIVIDE()` to handle division by 0 or NULL:
+--  https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#mathematical-functions
+--  Also considered z-score standardisation:
+--  https://www.investopedia.com/terms/z/zscore.asp#:~:text=A%20Z%2Dscore%20is%20a,identical%20to%20the%20mean%20score
+
+WITH cte_all AS
+(
+    SELECT
+        TIMESTAMP_SECONDS(ga.visitStartTime) AS visitStartTime
+        ,CAST(TIMESTAMP_SECONDS(ga.visitStartTime) AS DATE) AS visitStartDate
+        ,hits.page.pagePath
+        ,(SELECT value FROM hits.customDimensions WHERE index = 2) AS documentType
+        ,hits.page.searchKeyword AS searchKeyword
+        ,hits.page
+    FROM
+        `govuk-bigquery-analytics.87773428.ga_sessions_*` AS ga,
+        UNNEST(hits) AS hits
+    WHERE
+        _table_suffix > FORMAT_DATE('%Y%m%d', DATE_SUB(CURRENT_DATE(), INTERVAL 1 YEAR))
+            AND hits.type = "PAGE"
+),
+
+cte_id AS
+(
+    SELECT
+        visitStartDate
+        ,EXTRACT(WEEK FROM visitStartDate) AS visitStartDateWeek
+        ,CASE
+            WHEN documentType = 'simple_smart_answer' THEN CONCAT('smart answers - ', pagePath)
+            WHEN searchKeyword IS NOT NULL THEN CONCAT('search - ', pagePath)
+            ELSE CONCAT('other - ', pagePath)
+            END AS pageId
+        ,page
+    FROM cte_all
+),
+
+cte_week_counts AS
+(
+    SELECT
+        visitStartDateWeek
+        ,pageId
+        ,COUNT(page) AS pageHits
+    FROM cte_id
+    GROUP BY
+        visitStartDateWeek
+        ,pageId
+),
+
+cte_normalise AS
+(
+    SELECT
+        visitStartDateWeek
+        ,pageId
+        ,pageHits
+        ,MAX(pageHits) OVER(PARTITION BY visitStartDateWeek) AS pageHitsMax
+        ,MIN(pageHits) OVER(PARTITION BY visitStartDateWeek) AS pageHitsMin
```

Comment:
It should be the highest and lowest number of hits out of all pages per week. Thus, it would be more like:

date | dateWeek | page | pageHits | pageHitsMax | pageHitsMin |
| --- | --- | --- | --- | --- | --- |
| 2021-01-01 | 1 | a | *12* | 56 | 12 |
| 2021-01-02 | 1 | b | **56** | 56 | 12 |
| 2021-01-06 | 1 | c | 23 | 56 | 12 |

Reply:
Cool, thanks for the explanation. So, we get pageHitsNormalised for each page for each week - do we then get the average hits per week per page at the end?

Reply:
Not yet, but can add that in!